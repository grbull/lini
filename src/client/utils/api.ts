import Axios from 'axios';

import { AuthLoginDto, AuthValidateDto } from '../../server/auth/auth.dto';
import {
  EpisodeDto,
  EpisodeScheduleDto,
} from '../../server/episode/episode.dto';
import { PushSubscriptionCreateDto } from '../../server/push_subscription/push_subscription.dto';
import { ShowDto, ShowWithEpisodesDto } from '../../server/show/show.dto';
import {
  SubscriptionCreateDto,
  SubscriptionDto,
  SubscriptionRemoveDto,
  // SubscriptionUpdateDto,
} from '../../server/subscription/subscription.dto';
import { UserDto, UserUpdateDto } from '../../server/user/user.dto';

const axios = Axios.create({
  baseURL: process.env.SERVER_URL,
  withCredentials: true,
});

async function userLogin(loginDto: AuthLoginDto): Promise<void> {
  await axios.post<void>('auth/login', loginDto);
}

async function userValidate(validateDto: AuthValidateDto): Promise<UserDto> {
  const response = await axios.post<UserDto>('auth/validate', validateDto);
  return response.data;
}

async function userGet(): Promise<UserDto> {
  const response = await axios.get<UserDto>('user');
  return response.data;
}

async function userUpdate(updateDto: UserUpdateDto): Promise<UserDto> {
  const response = await axios.put<UserDto>('user', updateDto);
  return response.data;
}

async function showSearch(query: string): Promise<ShowDto[]> {
  const response = await axios.get<ShowDto[]>(`show/search/${query}`);
  return response.data;
}

async function showGet(id: number): Promise<ShowWithEpisodesDto> {
  const response = await axios.get<ShowWithEpisodesDto>(`show/${id}`);
  return response.data;
}

async function episodeGet(id: number): Promise<EpisodeDto> {
  const response = await axios.get<EpisodeDto>(`episode/${id}`);
  return response.data;
}

async function episodeGetByShow(id: number): Promise<EpisodeDto[]> {
  const response = await axios.get<EpisodeDto[]>(`episode/show/${id}`);
  return response.data;
}

async function subscriptionGetAll(): Promise<SubscriptionDto[]> {
  const response = await axios.get<SubscriptionDto[]>('subscription/all');
  return response.data;
}

async function subscriptionCreate(
  createDto: SubscriptionCreateDto
): Promise<SubscriptionDto> {
  const response = await axios.post<SubscriptionDto>('subscription', createDto);
  return response.data;
}

// async function subscriptionUpdate(
//   updateDto: SubscriptionUpdateDto
// ): Promise<SubscriptionDto> {
//   const response = await axios.put<SubscriptionDto>('subscription', updateDto);
//   return response.data;
// }

async function subscriptionRemove(
  removeDto: SubscriptionRemoveDto
): Promise<SubscriptionDto> {
  const response = await axios.delete<SubscriptionDto>('subscription', {
    data: removeDto,
  });
  return response.data;
}

async function episodeScheduleGet(): Promise<EpisodeScheduleDto> {
  const response = await axios.get<EpisodeScheduleDto>('episode/schedule');
  return response.data;
}

async function pushCreate(createDto: PushSubscriptionCreateDto): Promise<void> {
  await axios.post('push_subscription', createDto);
}

export const api = {
  user: {
    login: userLogin,
    validate: userValidate,
    get: userGet,
    update: userUpdate,
  },
  show: {
    search: showSearch,
    get: showGet,
  },
  episode: {
    get: episodeGet,
    getByShow: episodeGetByShow,
    getSchedule: episodeScheduleGet,
  },
  subscription: {
    getAll: subscriptionGetAll,
    create: subscriptionCreate,
    // update: subscriptionUpdate,
    remove: subscriptionRemove,
  },
  push: {
    create: pushCreate,
  },
};
