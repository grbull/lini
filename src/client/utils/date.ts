import { DateTime } from 'luxon';

export function dateFromNow(airstamp: string): string {
  const { locale } = Intl.DateTimeFormat().resolvedOptions();
  return DateTime.fromISO(airstamp, { locale }).toRelative() || airstamp;
}

export function dateToLocaleDate(airstamp: string): string {
  const { locale } = Intl.DateTimeFormat().resolvedOptions();
  return (
    DateTime.fromISO(airstamp, { locale }).toLocaleString(DateTime.DATE_FULL) ||
    airstamp
  );
}

export function dateToLocaleTime(airstamp: string): string {
  const { locale } = Intl.DateTimeFormat().resolvedOptions();
  return (
    DateTime.fromISO(airstamp, { locale }).toLocaleString(
      DateTime.TIME_WITH_SHORT_OFFSET
    ) || ''
  );
}
