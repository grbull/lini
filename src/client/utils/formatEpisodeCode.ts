export function formatEpisodeCode(season: number, episode: number): string {
  let s: string = season.toString(10);
  let e: string = episode.toString(10);
  if (season < 10) {
    s = `0${s}`;
  }
  if (episode < 10) {
    e = `0${e}`;
  }
  return `S${s}E${e}`;
}
