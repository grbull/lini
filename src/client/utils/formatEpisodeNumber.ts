export function formatEpisodeNumber(number: number): string {
  if (number < 10) {
    return `0${number.toString(10)}`;
  }
  return number.toString();
}
