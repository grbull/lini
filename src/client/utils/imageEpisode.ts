// TODO: Create placeholder
const placeHolder = '/placeholder_episode.png';

export function imageEpisode(url: string | null): string {
  if (url) {
    return process.env.IMAGE_PROXY_URL + url;
  }
  return placeHolder;
}
