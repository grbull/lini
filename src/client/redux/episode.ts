/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type EpisodeState = {
  isLoading: boolean;
  error?: string;
  data?: EpisodeDto;
};

const initialState: EpisodeState = {
  isLoading: false,
};

export const get = createAsyncThunk<EpisodeDto, number>('episode/get', (id) =>
  api.episode.get(id)
);

export const episode = createSlice({
  name: 'episode',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const episodeActions = { ...episode.actions, get };
