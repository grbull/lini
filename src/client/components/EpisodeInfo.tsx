import cn from 'classnames';
import React, { ReactElement } from 'react';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { dateToLocaleDate, dateToLocaleTime } from '../utils/date';
import { formatEpisodeCode } from '../utils/formatEpisodeCode';
import { imageEpisode } from '../utils/imageEpisode';
import { FixedImage } from './FixedImage';
import { InfoCard } from './InfoCard';

interface Props {
  episode: EpisodeDto;
}

export function EpisodeInfo({ episode }: Props): ReactElement {
  return (
    <InfoCard>
      <div className={cn('w-full', 'h-6', 'flex', 'justify-between')}>
        <h2 className={cn('font-bold')}>{episode.name}</h2>
        <p>{formatEpisodeCode(episode.season, episode.number)}</p>
      </div>
      <div className={cn('w-full', 'h-6', 'flex', 'justify-between')}>
        <p>{episode.airstamp ? dateToLocaleDate(episode.airstamp) : 'N/A'}</p>
        <p>{episode.airstamp ? dateToLocaleTime(episode.airstamp) : 'N/A'}</p>
      </div>
      <FixedImage
        alt={episode.name}
        className={cn('my-3')}
        fullHeight={1080}
        fullWidth={1920}
        src={imageEpisode(episode.imageOriginal)}
      />
      <p className={cn('min-h-16')}>
        {episode.summary || 'Summary unavailable.'}
      </p>
    </InfoCard>
  );
}
