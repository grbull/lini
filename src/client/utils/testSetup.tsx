import { configureStore } from '@reduxjs/toolkit';
import { render, RenderResult } from '@testing-library/react';
import { createMemoryHistory, MemoryHistory } from 'history';
import React, { ReactNode } from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';

import { reducers, RootState } from '../redux/store';

export interface TestSetupProps {
  history?: MemoryHistory;
  state?: Partial<RootState>;
}

export interface TestSetupResult extends RenderResult {
  dispatch: typeof jest.fn;
  history: MemoryHistory<unknown>;
  scrollTo: typeof jest.fn;
}

export function testSetup(
  children: ReactNode,
  props?: TestSetupProps
): TestSetupResult {
  const history = props?.history || createMemoryHistory();
  const store = configureStore({
    reducer: reducers,
    preloadedState: props?.state || undefined,
  });

  // const reduxHistory = [];
  // const dispatch = (action: any): any => reduxHistory.push(action);
  const dispatch = jest.fn();
  store.dispatch = dispatch;

  const scrollTo = jest.fn();
  window.scrollTo = scrollTo;

  const utils = render(
    <Router history={history}>
      <Provider store={store}>{children}</Provider>
    </Router>
  );

  return { dispatch, history, scrollTo, ...utils };
}
