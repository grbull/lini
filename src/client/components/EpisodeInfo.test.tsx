/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import React from 'react';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { testSetup } from '../utils/testSetup';
import { EpisodeInfo } from './EpisodeInfo';

const episode: EpisodeDto = {
  id: '1627015',
  name: 'Red Moon',
  season: 1,
  number: 1,
  type: 'regular',
  airdate: '2019-11-01',
  airtime: null,
  airstamp: '2019-11-01T12:00:00.000Z',
  runtime: 65,
  imageMedium: '/uploads/images/medium_landscape/218/545490.jpg',
  imageOriginal: '/uploads/images/original_untouched/218/545490.jpg',
  summary:
    'NASA is in crisis as the Soviets land the first man on the moon in 1969, the beginning of an alternate history.',
  show: {
    id: 41414,
    name: 'For All Mankind',
    type: 'Scripted',
    language: 'English',
    genres: ['Drama', 'Science-Fiction'],
    status: 'Running',
    runtime: null,
    runtimeAverage: 61,
    datePremiered: '2019-11-01',
    officialSite:
      'https://tv.apple.com/us/show/for-all-mankind/umc.cmc.6wsi780sz5tdbqcf11k76mkp7',
    scheduleTime: null,
    scheduleDays: ['Friday'],
    rating: 7.1,
    imageMedium: '/uploads/images/medium_portrait/293/732962.jpg',
    imageOriginal: '/uploads/images/original_untouched/293/732962.jpg',
    summary:
      'Imagine a world where the global space race never ended. This "what if" take on history from Ronald D. Moore (Outlander, Battlestar Galactica) spotlights the lives of NASA astronauts—the heroes and rock stars of their time—and their families.',
    network: null,
    webChannel: null,
  },
};

describe('EpisodeInfo Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = testSetup(<EpisodeInfo episode={episode} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render episodes without an airstamp', () => {
    const { getAllByText } = testSetup(
      <EpisodeInfo episode={{ ...episode, airstamp: null }} />
    );

    expect(getAllByText('N/A')).toHaveLength(2);
  });

  it('should render episodes without a summary', () => {
    const { getByText } = testSetup(
      <EpisodeInfo episode={{ ...episode, summary: null }} />
    );

    expect(getByText('Summary unavailable.')).toBeTruthy();
  });
});
