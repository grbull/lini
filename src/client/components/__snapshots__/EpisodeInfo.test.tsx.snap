// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodeInfo Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="m-2 px-2 py-3 bg-gray-200 dark:bg-gray-900 bg-opacity-90 dark:bg-opacity-80 overflow-hidden"
  >
    <a
      class="hidden md:block text-xl text-center mb-4"
      href="/show/41414"
    >
      For All Mankind
    </a>
    <div
      class="w-full h-6 flex justify-between"
    >
      <h2
        class="font-bold"
      >
        Red Moon
      </h2>
      <p>
        S01E01
      </p>
    </div>
    <div
      class="w-full h-6 flex justify-between"
    >
      <p>
        November 1, 2019
      </p>
      <p>
        12:00:00 PM GMT
      </p>
    </div>
    <div
      class="h-0 relative my-3"
      style="padding-top: 56.25%;"
    >
      <div
        class="animate-pulse bg-gray-700 absolute top-0 left-0 w-full h-full"
      />
      <img
        alt="Red Moon"
        class="absolute top-0 left-0 w-full h-full hidden"
        src="undefined/uploads/images/original_untouched/218/545490.jpg"
      />
    </div>
    <p
      class="min-h-16"
    >
      NASA is in crisis as the Soviets land the first man on the moon in 1969, the beginning of an alternate history.
    </p>
  </div>
</DocumentFragment>
`;
